pkg/asset/manifests/gcp/cluster.go

```diff
@@ -0,0 +1,171 @@
+	}
+
+	worker := capg.SubnetSpec{
+		Name:        workerSubnet,
+		CidrBlock:   "",
+		Description: ptr.To(description),
+		Region:      installConfig.Config.GCP.Region,
+	}
+
+	// Add the CIDR information.
+	machineV4CIDRs := []string{}
+	for _, network := range installConfig.Config.Networking.MachineNetwork {
+		if network.CIDR.IPNet.IP.To4() != nil {
+			machineV4CIDRs = append(machineV4CIDRs, network.CIDR.IPNet.String())
+		}
+	}
+
+	if len(machineV4CIDRs) == 0 {
+		return nil, fmt.Errorf("failed to parse machine CIDRs")
+	}
+
+	_, ipv4Net, err := net.ParseCIDR(machineV4CIDRs[0])
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse machine network CIDR: %w", err)
+	}
+
+	if installConfig.Config.GCP.ControlPlaneSubnet == "" {
+		masterCIDR, err := cidr.Subnet(ipv4Net, 1, 0)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create the master subnet %w", err)
+		}
+		master.CidrBlock = masterCIDR.String()
+	}
+
+	if installConfig.Config.GCP.ComputeSubnet == "" {
+		workerCIDR, err := cidr.Subnet(ipv4Net, 1, 1)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create the worker subnet %w", err)
+		}
+		worker.CidrBlock = workerCIDR.String()
+	}
+
+	subnets := []capg.SubnetSpec{master, worker}
+
+	labels := map[string]string{}
+	labels[fmt.Sprintf("kubernetes-io-cluster-%s", clusterID.InfraID)] = "owned"
+	labels[fmt.Sprintf("capg-cluster-%s", clusterID.InfraID)] = "owned"
+	for _, label := range installConfig.Config.GCP.UserLabels {
+		labels[label.Key] = label.Value
+	}
+
+	// Find the availability zones and use this information to configure the GCP Cluster Failure Domains.
+	zones := sets.New[string]()
+	defaultMachinePlatformDefined := false
+	if installConfig.Config.Platform.GCP != nil {
+		if installConfig.Config.Platform.GCP.DefaultMachinePlatform != nil {
+			defaultMachinePlatformDefined = true
+			for _, zone := range installConfig.Config.Platform.GCP.DefaultMachinePlatform.Zones {
+				zones.Insert(zone)
+			}
+
+			// The default machine platform is defined take any zones from the compute node(s) and
+			// any defined in the control plane node(s).
+			// Note that this is different from when the default machine platform is not defined,
+			// where the zones are only used if both the compute and control plane node information exists.
+			if installConfig.Config.ControlPlane.Platform.GCP != nil {
+				for _, zone := range installConfig.Config.ControlPlane.Platform.GCP.Zones {
+					zones.Insert(zone)
+				}
+			}
+			if installConfig.Config.Compute[0].Platform.GCP != nil {
+				for _, zone := range installConfig.Config.Compute[0].Platform.GCP.Zones {
+					zones.Insert(zone)
+				}
+			}
+		}
+	}
+	if !defaultMachinePlatformDefined {
+		// when the default platform is not defined, then check that both compute and control plane node information
+		// exists before adding any data.
+		if installConfig.Config.ControlPlane.Platform.GCP != nil && installConfig.Config.Compute[0].Platform.GCP != nil {
```

Comment:
We should only do this if the zones on the machinepool are populated, so these checks should be for `len` of the zones, not just that the machine pools exist.